### ISSUE TITLE
Failed to get terraform version

### ISSUE THREAD

**Issue by takayamaki** [hashicorp/terraform-ls#151]
Language server has failed to get terraform version.
It refers ~/.tfenv/version to get version.
But I specify version by .terraform-version file  in top level of repository.

### Server Version
```
0.3.2
```

### Terraform Version
```
0.12.20
```

### Client Version
```
Terraform VSCode Extention v2.0.0-rc1
```

### Terraform Configuration Files
omit

### Log Output
```
2020/06/11 00:13:41 exec.go:176: Starting /home/my_user_name/.tfenv/bin/terraform ["terraform" "version"] in "/tmp"...
2020/06/11 00:13:41 exec.go:142: Waiting for command to finish ...
2020/06/11 00:13:48 rpc_logger.go:45: Error for "initialize" (ID 0): [-32098] failed to get version: terraform (pid 14318) exited (code 1): exit status 1
stdout: ""
stderr: "cat: /home/my_user_name/.tfenv/version: No such file or directory\n\x1b[31mVersion could not be resolved (set by /home/my_user_name/.tfenv/version or tfenv use <version>)\x1b[0m\n"
2020/06/11 00:13:48 server.go:204: Completed 1 requests [7.7687106s elapsed]
[Error - 12:13:48 AM] Server initialization failed.
  Message: failed to get version: terraform (pid 14318) exited (code 1): exit status 1
stdout: ""
stderr: "cat: /home/my_user_name/.tfenv/version: No such file or directory\n\x1b[31mVersion could not be resolved (set by /home/my_user_name/.tfenv/version or tfenv use <version>)\x1b[0m\n"
  Code: -32098 
2020/06/11 00:13:48 server.go:469: Received 1 new requests
2020/06/11 00:13:48 server.go:165: Processing 1 requests
2020/06/11 00:13:48 server.go:223: Checking request for "shutdown": 
2020/06/11 00:13:48 rpc_logger.go:29: Incoming request for "shutdown" (ID 1): 
2020/06/11 00:13:48 rpc_logger.go:50: Response to "shutdown" (ID 1): null
2020/06/11 00:13:48 server.go:204: Completed 1 requests [586.8µs elapsed]
2020/06/11 00:13:48 server.go:469: Received 1 new requests
2020/06/11 00:13:48 server.go:165: Processing 1 requests
2020/06/11 00:13:48 server.go:223: Checking request for "exit": 
2020/06/11 00:13:48 rpc_logger.go:29: Incoming notification for "exit":
```

### Expected Behavior
Language server refers .terraform-version file.

### Actual Behavior
Language server failed initialize.

### Steps to Reproduce
1. `rm ~/.tfenv/version`
1. `echo 0.12.20 > .terraform-version` in top level of repository
1. open repository by VSCode with terraform extention

**Comment by radeksimko** [comment]
I was reading through the relevant code paths and toying with tfenv a bit recently.

Even if/when the language server is able to detect the existence of tfenv (i.e. whether `.terraform-version` exists in PWD and `tfenv` in `$PATH`) the tricky thing is that we make some (IMO sane) assumptions about what certain commands will do and what they'll output.

For example we assume that `terraform version` will output version to STDOUT, which may not be true in the scenario @takayamaki outlined above:

```
➜  github echo 0.12.20 > .terraform-version
➜  github terraform version
version '0.12.20' is not installed (set by /var/workspace/tf-test/github/.terraform-version). Installing now as TFENV_AUTO_INSTALL==true
Installing Terraform v0.12.20
Downloading release tarball from https://releases.hashicorp.com/terraform/0.12.20/terraform_0.12.20_darwin_amd64.zip
################################################################################################################################## 100.0%
Downloading SHA hash file from https://releases.hashicorp.com/terraform/0.12.20/terraform_0.12.20_SHA256SUMS
▶ ERROR No UID given but one was expected
Unable to verify OpenPGP signature unless logged into keybase and following hashicorp
Archive:  tfenv_download.Gnbe7m/terraform_0.12.20_darwin_amd64.zip
  inflating: /usr/local/Cellar/tfenv/2.0.0/versions/0.12.20/terraform
Installation of terraform v0.12.20 successful. To make this your default version, run 'tfenv use 0.12.20'
Terraform v0.12.20
+ provider.github v2.8.1

Your version of Terraform is out of date! The latest version
is 0.12.28. You can update by downloading from https://www.terraform.io/downloads.html
```

^
All of the above except the `#` progress is sent to STDOUT unfortunately, which then makes it nearly impossible to parse the output.

Also this kind of behaviour is likely to break any machine-readable JSON output, because the "progress report" just isn't valid JSON.

One potential solution might be for `tfenv` to send tfenv-specific output to STDERR, rather STDOUT and reserve STDOUT from `terraform` commands to Terraform. That said I am not sure yet how we may represent machine-readable errors and whether these would also be sent to STDERR.

Another solution would be deeper integration with tfenv where we could turn tfenv into strict mode where it isn't installing anything automatically, but rather tell us whether the version is installed and if not, we can just flag it up early to the user,  prompt them for the installation, or even install it automatically. These steps however need to very much more explicitly defined.

**Comment by joshbranham** [comment]
I believe I am also hitting this using `asdf`, my only reasoning is whenever I launch VSCode with this enabled, I end up with ~100 processes of `terraform version`. Unsure if that number grows or not, as my laptop completely locks up at that point. 

For example:
```
❯ ps aux | grep terraform
joshuabranham    21761  12.6  0.1  4433728  15196   ??  R     2:51PM   0:06.32 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    14819   3.5  0.1  4432384  14352   ??  U     2:50PM   0:07.72 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    18355   3.5  0.1  4430592  14480   ??  U     2:50PM   0:06.71 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    20057   3.3  0.1  4432128  14408   ??  U     2:50PM   0:06.83 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    13350   3.3  0.1  4431616  14360   ??  U     2:50PM   0:08.06 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    21701   3.3  0.1  4430516  15172   ??  R     2:51PM   0:06.34 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    23520   3.3  0.1  4430592  14944   ??  R     2:51PM   0:06.29 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    19248   3.2  0.1  4431360  14420   ??  U     2:50PM   0:06.65 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    23627   3.2  0.1  4435712  16576   ??  R     2:51PM   0:06.08 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    19860   3.2  0.1  4428928  14332   ??  U     2:50PM   0:06.62 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    19254   3.2  0.1  4431616  14496   ??  U     2:50PM   0:06.89 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    19250   3.2  0.1  4431424  14364   ??  R     2:50PM   0:06.61 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    20751   3.2  0.1  4436992  14400   ??  U     2:51PM   0:06.76 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    10535   3.2  0.1  4430336  14284   ??  R     2:50PM   0:08.48 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham     9590   3.2  0.1  4432640  14364   ??  R     2:50PM   0:08.56 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    15806   3.2  0.1  4430592  14440   ??  U     2:50PM   0:07.37 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    15837   3.2  0.1  4430336  14268   ??  R     2:50PM   0:07.53 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham     9584   3.1  0.1  4436992  14464   ??  R     2:50PM   0:08.26 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    22331   3.1  0.1  4434688  14828   ??  U     2:51PM   0:06.18 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    16912   3.1  0.1  4430592  14400   ??  U     2:50PM   0:07.57 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    21913   3.1  0.1  4433408  14912   ??  U     2:51PM   0:06.41 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    15400   3.1  0.1  4430592  14376   ??  R     2:50PM   0:07.66 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    18358   3.1  0.1  4430592  14464   ??  U     2:50PM   0:06.64 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    12957   3.1  0.1  4431616  14392   ??  U     2:50PM   0:08.04 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    13427   3.1  0.1  4433472  14344   ??  U     2:50PM   0:07.76 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    12418   3.1  0.1  4430592  14256   ??  R     2:50PM   0:07.97 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    16753   3.1  0.1  4429184  14292   ??  U     2:50PM   0:07.33 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    12638   3.1  0.1  4431616  14568   ??  U     2:50PM   0:07.98 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    21488   3.1  0.1  4431616  14772   ??  U     2:51PM   0:06.45 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    21228   3.1  0.1  4430592  14900   ??  U     2:51PM   0:06.39 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    23628   3.1  0.1  4436736  16680   ??  R     2:51PM   0:05.89 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    12038   3.1  0.1  4430336  14304   ??  R     2:50PM   0:08.23 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    12619   3.1  0.1  4433408  14436   ??  U     2:50PM   0:07.95 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    13429   3.1  0.1  4431744  14336   ??  R     2:50PM   0:07.77 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    12670   3.0  0.1  4431872  14416   ??  U     2:50PM   0:08.09 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham     9598   3.0  0.1  4433664  14420   ??  R     2:50PM   0:08.22 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham     9593   3.0  0.1  4434304  14236   ??  R     2:50PM   0:08.52 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham     9582   3.0  0.1  4434792  14196   ??  U     2:50PM   0:08.64 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    23209   3.0  0.1  4433920  14420   ??  U     2:51PM   0:06.01 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    22859   3.0  0.1  4435456  15336   ??  R     2:51PM   0:06.33 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    22665   3.0  0.1  4436736  15344   ??  U     2:51PM   0:06.10 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    22664   3.0  0.1  4431616  15144   ??  U     2:51PM   0:06.27 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    21669   3.0  0.1  4434764  14588   ??  U     2:51PM   0:06.48 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
joshuabranham    21604   3.0  0.1  4434432  14356   ??  U     2:51PM   0:06.51 /Users/joshuabranham/.asdf/installs/terraform/0.12.26/bin/terraform version
```

**Comment by radeksimko** [comment]
@joshbranham That sounds quite accurate - since `v0.5` the language server issues `terraform version` for each found root module, which is exactly to reflect the reality of people opening deep hierarchy of different modules with different Terraform versions. It's still not ideal in case of `tfenv` though for the reasons I mentioned above, where it auto-installs missing versions and pollutes stdout. That said in theory if all versions get installed once, then the discovery on subsequent IDE launch should work just fine.

The root cause behind so many processes may also be https://github.com/tfutils/tfenv/issues/196 - 1 second of delay can make a big difference for 50 processes - they would still get executed without tfenv, but the execution overall would probably take just a second or two for all of them, instead of a minute and less processes would then compete for the CPU at the same time.

Admittedly if each of these 50 root modules are also initialized the server would also run `terraform providers schema -json` which can take a few seconds on bigger providers even without tfenv (and consume CPU/memory). So I think we do need to look into the scalability of this model and/or cap the number of root modules which get preloaded automatically to avoid locking up CPU on larger hierarchies. That problem is tracked under https://github.com/hashicorp/terraform-ls/issues/186

**Comment by joshbranham** [comment]
> @joshbranham That sounds quite accurate - since `v0.5` the language server issues `terraform version` for each found root module, which is exactly to reflect the reality of people opening deep hierarchy of different modules with different Terraform versions. It's still not ideal in case of `tfenv` though for the reasons I mentioned above, where it auto-installs missing versions and pollutes stdout. That said in theory if all versions get installed once, then the discovery on subsequent IDE launch should work just fine.
> 
> The root cause behind so many processes may also be [tfutils/tfenv#196](https://github.com/tfutils/tfenv/issues/196) - 1 second of delay can make a big difference for 50 processes - they would still get executed without tfenv, but the execution overall would probably take just a second or two for all of them, instead of a minute and less processes would then compete for the CPU at the same time.
> 
> Admittedly if each of these 50 root modules are also initialized the server would also run `terraform providers schema -json` which can take a few seconds on bigger providers even without tfenv (and consume CPU/memory). So I think we do need to look into the scalability of this model and/or cap the number of root modules which get preloaded automatically to avoid locking up CPU on larger hierarchies. That problem is tracked under #186

Hey thanks for the quick response! In our case its all `>= 0.12` for the tf version, and _most_ of the directories are not `terraform init` since we use a git bot for most of our plan/applies. I will work around it for now and just open the specific module folder in my editor

**Comment by radeksimko** [comment]
The CPU/memory consumption issue was resolved as part of the linked issue and some other following PRs.

As for the originally reported problem where the language server would fail to get Terraform version. Given the context I described in one of my above comments I would say that the language behaves as intended, given that `terraform version` doesn't provide the expected output.

We did discuss integrating tfenv within the team, but I'm currently unsure of the ideal UX around it, so if you do expect/want tighter tfenv integration I would suggest joining the conversation and sharing your thoughts here: https://github.com/hashicorp/terraform-ls/issues/305
